enhancements/ocp-coreos-layering.md

```diff
@@ -0,0 +1,485 @@
+- [ ] Gain [transactional configuration+code changes](https://github.com/openshift/machine-config-operator/issues/1190) by having configuration+encode captured atomically in a container
+- [ ] Avoid breaking existing workflow via MachineConfig (including extensions)
+- [ ] Avoid overwriting existing custom modifications (such as files managed by other operators) during upgrades
+
+### Non-Goals
+
+- While the base CoreOS layer/ostree-container tools will be usable outside of OpenShift, this enhancement does not cover or propose any in-cluster functionality for exporting the forward-generated image outside of an OpenShift cluster. In other words, it is not intended to be booted (`$ rpm-ostree rebase <image>`) from outside of a cluster.
+- This proposal does not cover generating updated "bootimages"; see https://github.com/openshift/enhancements/pull/201
+- Don't change existing workflow for RHEL worker nodes
+
+## Proposal
+
+**NOTE: Nothing in this proposal should be viewed as final.  It is highly likely that details will change.  It is quite possible that larger architectural changes will be made as well.**
+
+1. The `machine-os-content` shipped as part of the release payload will change format to the new "native ostree-container" format, in which the OS content appears as any other OCI/Docker container.  
+  (In contrast today, the existing `machine-os-content` has an ostree repository inside a UBI image, which is hard to inspect and cannot be used for derived builds).  For more information, see [ostree-rs-ext](https://github.com/ostreedev/ostree-rs-ext/) and [CoreOS layering](https://github.com/coreos/enhancements/pull/7). 
+2. Documentation and tooling will be available for generating derived images from this base image
+3. This tooling will be used by the MCO to create derived images (more on this in a separate enhancement)
+
+
+### User Stories
+
+#### What works now continues to work
+
+An OpenShift administrator at example.corp is happily using OpenShift 4 (with RHEL CoreOS) in several AWS clusters today, and has only a small custom MachineConfig object to tweak host level auditing.  They do not plan to use any complex derived builds, and expect that upgrading their existing cluster continues to work and respect their small audit configuration change with no workflow changes.
+
+#### Adding a 3rd party security scanner/IDS
+
+example.bank's security team requires a 3rd party security agent to be installed on bare metal machines in their datacenter.  The 3rd party agent comes as an RPM today, and requires its own custom configuration.  While the 3rd party vendor has support for execution as a privileged daemonset on their roadmap, it is not going to appear soon. 
+
+After initial cluster provisioning is complete, the administrators at example.bank supply a configuration that adds a repo file to `/etc/yum.repos.d/agentvendor.repo` and requests installation of a package named `some-3rdparty-security-agent` as part of a container build.
```

Comment:
is it also possible to do it on an initial cluster install? by just doing an `openshift-install create manifests` and then adding this custom manifest with this config which MCO would rollout during install?

Reply:
We could clearly use the bootstrap node to perform a build (e.g. `podman build`).  I think performance needs will drive us to do this because otherwise we'd be adding a whole other phase to rollout where we'd need to wait for the cluster to be up to do a build, to generate our container image to roll it out to all the nodes.

Reply:
I suspect a number of administrators though will end up wanting to provide an *externally* built image; that's alluded to above.  A whole host of tradeoffs with thaat.

Reply:
hmm..in that case where would the builds be pushed though ? so that it can be picked up my the individual nodes' MCO?

Reply:
We'd run a registry on the bootstrap node.

Reply:
rigth..externally built images is definitely an option but most of the time we have requests like - "i want to add this systemd unti/file to ignition " or just change some kargs for which this will be useful

Reply:
Yep definitely, all that works via MachineConfig today and we intend to keep it working.